### PR TITLE
update to simplified ecal pfcluster corrections

### PR DESCRIPTION
**Pull request by bendavid**
Move to updated version of ecal pfcluster corrections with reduced number of input variables, plus some additional pt binning and slightly different parameterization to improve the behaviour at very low energies.

Having some problems reading the conditions from the db at the moment, so the branch currently is pointing to an sqlite file on my public afs area (clearly the minimum to merge the pull request would be to have the esprefer working from FrontierProd, if not updated global tags)

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

File: RecoParticleFlow/PFClusterProducer/python/particleFlowClusterECAL_cfi.py [hunk 1]
```diff
@@ -1,5 +1,265 @@
...
+)
+
+#pfcluscorsource.connect = cms.string('frontier://FrontierProd/CMS_COND_PAT_000')
+pfcluscorsource.connect = cms.string("sqlite_file:/afs/cern.ch/user/b/bendavid/work/public/ecalPFClusterCorV2_22042015.db")
+
+pfclusprefer = cms.ESPrefer(
+    'PoolDBESSource',
+    'pfcluscorsource',
+    GBRDWrapperRcd = cms.vstring(
+                                'GBRForestD/ecalPFClusterCorV2_EB_pfSize1_mean_50ns',
```
Comment (slava77) [inline RecoParticleFlow/PFClusterProducer/python/particleFlowClusterECAL_cfi.py]: so much for a light training, 20 records to get the data out.
For the future, is there a way for GBRForest to support categories from just one object?

File: RecoParticleFlow/PFClusterProducer/src/PFClusterEMEnergyCorrector.cc [hunk 2]
```diff
@@ -133,42 +148,32 @@ void PFClusterEMEnergyCorrector::correctEnergies(const edm::Event &evt, const ed
...
     es.get<GBRDWrapperRcd>().get(condnames_sigma[icor],forestH_sigma[icor]);
   }
   
-  std::array<float,11> eval;
+  std::array<float,5> eval;
     
   EcalClusterLazyTools lazyTool(evt, es, _recHitsEB, _recHitsEE);
-
-  //count number of primary vertices for pileup correction
-  edm::Handle<reco::VertexCollection> vtxH;
```
Comment (slava77) [inline RecoParticleFlow/PFClusterProducer/src/PFClusterEMEnergyCorrector.cc]: is there no PU dependence anymore?

**Comment by bendavid** [comment]
@konush @lgray @matteosan1

Slava, GBRForest does not/will not support categorization itself, but it might be possible for the future to create a wrapper object which does that.  (Actually a more ambitious project as @davidlange6 was suggesting in the ORP yesterday would be to use some kind of string parsing, etc to make the conditions objects also define at some level which variables are being used and in what order so they could be filled semi-automatically)

For the addition of the pt-binnings expanding the number of trainings, this would not have been my first choice, but bottom line is that this was done in a rush in reaction to the jet met findings, and also because the gun samples, despite having been previously used, had been trapped on tape for the last several weeks.  This is the best/safest version of the training which is available at this exact moment.  @konush is still working on improvements/refinements, but since I think we cannot promise anything on very short timescale, we want to go ahead with what we have.  The training is still "light" in terms of the number of input variables.

Indeed this version of the training does not include any pileup dependence, since it was trained on the no pileup samples.  The reason is that there are some not fully understood/desirable features in the trainings on pileup samples when the cluster shapes are dropped.  Most likely the underlying cause is that the slope of the response vs pileup changes with the size of the cluster (known effect since a long time), but without the cluster shapes in the training, the MVA cannot compensate for that, and some types of clusters may even have a negative slope of corrected response vs pileup (bad for jetmet corrections among other things)

**Comment by bendavid** [comment]
(switching to the correct connect string so that for now at least the ESPrefer is reading from the db instead of sqlite file on afs)

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
(no pileup) dijet flat spectrum up to 3 TeV:
- increase (over threshold?) of the particleFlowSuperClusterECAL in barrel and endcap in the range with ET<10 GeV (E up to 30)
all_sign542vsorig-slc6-gcc491_qcd13tevflatpt15s3000wf1338p0c_log10recocaloclusters_particleflowsuperclusterecal_particleflowbasicclusterecalbarrel_reco_obj_energy
all_sign542vsorig-slc6-gcc491_qcd13tevflatpt15s3000wf1338p0c_log10recocaloclusters_particleflowsuperclusterecal_particleflowbasicclusterecalendcap_reco_obj_energy
The R9 of the cluster shapes changes quite a bit (almost nothing above 1 anymore; this looks more like it was before the MVA corrections)
wf1338_eb_sc_r9
wf1338_ee_sc_r9

- this propagates to egamma candidates
all_sign542vsorig-slc6-gcc491_qcd13tevflatpt15s3000wf1338p0c_log10recopfcandidates_particleflowegamma__reco_obj_pt
- and eventually to pfparticles, mainly visible as a shift in photon pt spectrum 
all_sign542vsorig-slc6-gcc491_qcd13tevflatpt15s3000wf1338p0c_log10recopfcandidates_particleflow__reco_obj_pt40

MET looks OK 
all_sign542vsorig-slc6-gcc491_qcd13tevflatpt15s3000wf1338p0c_log10recopfmets_pfmet__reco_obj_pt

added electrons are mostly in the "background-like" (badtrack) electrons
all_sign542vsorig-slc6-gcc491_qcd13tevflatpt15s3000wf1338p0c_recogsfelectrons_gedgsfelectrons__reco_obj_classification

corrected jets get a mild (under a few %) increase in response
wf1338_e_ak4pf_respcor
wf1338_b_ak4pf_respcor

The jet resolution isn't changing much
wf1338_b_ak4pf200to600_respcor

Higher stat point from dijet 3-TeV doesn't show much change in jet response on the high end
wf1313_b_ak4pf_resp


Now for real electrons:
dielectron 1 TeV
slight increase in response here as well
all_sign542vsorig-slc6-gcc491_singleelectron13pt1000wf1316p0c_recogsfelectrons_gedgsfelectrons__reco_obj_pt
r9 is slimmer 
wf1316_e_sc_r9

dielectron 35 GeV slight increase in response as well
all_sign542vsorig-slc6-gcc491_singleelectron13pt35wf1317p0c_recopfcandidates_particleflow__reco_obj_pt19
similar change in the gedPhoton as well (no track contribution)
all_sign542vsorig-slc6-gcc491_singleelectron13pt35wf1317p0c_recophotons_gedphotons__reco_obj_et

the lower end of spectrum ( photons ET 10 GeV gun) is almost unchanged, although looks a bit broader
all_sign542vsorig-slc6-gcc491_singlegamma13pt10wf1318p0c_recophotons_gedphotons__reco_obj_et


ttbar pileup ~20 (run1)  with 50 ns
shows somewhat more significant effect in photon distribution than in no-PU
all_sign542vsorig-slc6-gcc491_ttbarpuwf202p0c_log10recopfcandidates_particleflow__reco_obj_pt40
photon ET fraction becomes more visible in the MET (this wasn't as visible in no-PU case)
all_sign542vsorig-slc6-gcc491_ttbarpuwf202p0c_recopfmets_pfmet__reco_obj_photonetfraction


PU35 at 25 ns appears to have moderately smaller effect on pf photons
all_sign542vsorig-slc6-gcc491_ttbarpuwf25202p0c_log10recopfcandidates_particleflow__reco_obj_pt40

jet response has a fairly small change here.

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/8818	5717860

tested with extended stat tests in CMSSW_7_4_0 /test area sign542/
changes in the physics response appear to be generally as expected (see plots earlier in this thread).

**Comment by davidlange6** [comment]
@bendavid - could you take out the ESprefer and add in PR8869 to this PR?

**Comment by slava77** [comment]
-1

superseded by #8872

**Pull request closed without merging**